middleware/grpc_stats.go

```diff
@@ -65,10 +84,85 @@ func (g *grpcStatsHandler) HandleRPC(ctx context.Context, rpcStats stats.RPCStat
 	}
 }
 
-func (g *grpcStatsHandler) TagConn(ctx context.Context, _ *stats.ConnTagInfo) context.Context {
-	return ctx
+func (g *grpcStatsHandler) TagConn(ctx context.Context, conn *stats.ConnTagInfo) context.Context {
+	return context.WithValue(ctx, contextKeyConnID, conn.LocalAddr.String()+":"+conn.RemoteAddr.String())
 }
 
 func (g *grpcStatsHandler) HandleConn(_ context.Context, _ stats.ConnStats) {
 	// Not interested.
 }
+
+// streamEntry represents a connection and its stream count.
+type streamEntry struct {
+	streamCount int
+	connID      string
+}
+
+// Less defines the sort order: descending by streamCount, then ascending by connID.
+func (a streamEntry) Less(b btree.Item) bool {
+	bb := b.(streamEntry)
+	if a.streamCount != bb.streamCount {
+		return a.streamCount > bb.streamCount // descending order
+	}
+	return a.connID < bb.connID
+}
+
+// StreamTracker tracks the number of streams per connection and the max.
+type StreamTracker struct {
+	mu      sync.RWMutex
+	tree    *btree.BTree           // sorted set of stream counts
+	connMap map[string]streamEntry // connID -> entry
+}
+
+func NewStreamTracker() *StreamTracker {
+	return &StreamTracker{
+		tree:    btree.New(2), // degree 2 is fine for small datasets
+		connMap: make(map[string]streamEntry),
+	}
+}
+
+func (st *StreamTracker) OpenStream(connID string) {
```

Comment:
Could you quickly benchmark `OpenStream()` and `CloseStream()` to see how long it takes? I just want to make sure we're not introducing a scaling limit with this global lock (which effectively serialise the opening / closing of streams). I think it's pretty quick, so nothing to worry about, but measuring it in a benchmark may be worth it.